File: shared/py/redisManager.py
"""
Redis Manager for FastAPI applications
Provides Redis connection and operations
"""

import asyncio
import redis.asyncio as redis
import os
import logging
from typing import Optional, Any, Dict, List
from contextlib import asynccontextmanager


class RedisManager:
    """Redis connection manager"""

    def __init__(self):
        self.client: Optional[redis.Redis] = None
        self.logger = logging.getLogger(__name__)
        self._loop: Optional[Any] = None  # Track event loop for reconnection

        # Redis configuration
        self.host = os.getenv("REDIS_HOST", "localhost")
        self.port = int(os.getenv("REDIS_PORT", "6379"))
        self.password = os.getenv("REDIS_PASSWORD")
        self.db = int(os.getenv("REDIS_DB", "0"))

    async def connect(self) -> None:
        """Connect to Redis (handles event loop changes)"""
        try:
            current_loop = asyncio.get_running_loop()
            
            # If event loop changed, close old client and create new one
            if self._loop is not None and self._loop is not current_loop:
                self.logger.debug("Event loop changed, creating new Redis client")
                if self.client:
                    try:
                        await self.client.close()
                    except Exception:
                        pass
                    self.client = None
            
            # Create new client if needed
            if self.client is None:
                self.client = redis.Redis(
                    host=self.host,
                    port=self.port,
                    password=self.password,
                    db=self.db,
                    decode_responses=True,
                )

                # Test connection
                await self.client.ping()
                self._loop = current_loop
                self.logger.info(f"Connected to Redis at {self.host}:{self.port}")

        except Exception as e:
            self.logger.error(f"Failed to connect to Redis: {e}")
            raise

    async def disconnect(self) -> None:
        """Disconnect from Redis"""
        if self.client:
            await self.client.close()
            self.logger.info("Disconnected from Redis")

    async def get(self, key: str) -> Optional[str]:
        """Get value from Redis"""
        if not self.client:
            return None
        return await self.client.get(key)

    async def set(self, key: str, value: str, expire: Optional[int] = None) -> bool:
        """Set value in Redis"""
        if not self.client:
            return False
        return await self.client.set(key, value, ex=expire)

    async def delete(self, key: str) -> int:
        """Delete key from Redis"""
        if not self.client:
            return 0
        return await self.client.delete(key)

    async def exists(self, key: str) -> bool:
        """Check if key exists in Redis"""
        if not self.client:
            return False
        return await self.client.exists(key)

    async def expire(self, key: str, time: int) -> bool:
        """Set expiration time for key"""
        if not self.client:
            return False
        return await self.client.expire(key, time)

    async def publish(self, channel: str, message: str) -> int:
        """Publish message to Redis channel"""
        if not self.client:
            return 0
        return await self.client.publish(channel, message)

    async def subscribe(self, *channels: str):
        """Subscribe to Redis channels"""
        if not self.client:
            return None
        pubsub = self.client.pubsub()
        await pubsub.subscribe(*channels)
        return pubsub

    @asynccontextmanager
    async def pipeline(self):
        """Redis pipeline context manager"""
        if not self.client:
            raise RuntimeError("Redis client not connected")

        async with self.client.pipeline() as pipe:
            yield pipe

    async def health_check(self) -> bool:
        """Check Redis connection health"""
        try:
            if not self.client:
                return False
            await self.client.ping()
            return True
        except Exception:
            return False

    async def scan_keys(self, pattern: str = "*", count: int = 100) -> List[str]:
        """
        Scan Redis keys matching a pattern.
        
        Args:
            pattern: Key pattern (e.g., "prefix:*")
            count: Number of keys to scan per iteration
            
        Returns:
            List of matching keys
        """
        if not self.client:
            return []
        
        keys = []
        cursor = 0
        while True:
            cursor, batch = await self.client.scan(cursor, match=pattern, count=count)
            keys.extend(batch)
            if cursor == 0:
                break
        return keys

File: shared/js/redisManager.ts
import IORedis from "ioredis";

/**
 * Redis Manager - Centralized Redis Connection Management
 *
 * Provides a singleton Redis connection instance across all services.
 * Similar to DatabaseManager for MongoDB, this ensures all Redis operations
 * (queues, caching, etc.) use the same connection pool.
 *
 * Features:
 * - Singleton pattern for connection reuse
 * - Auto-reconnection on failures
 * - Event logging for debugging
 * - Environment-based configuration
 *
 * Usage:
 * const redis = RedisManager.getInstance();
 * await redis.connect();
 * const value = await redis.get('key');
 */

export class RedisManager {
  private static instance: RedisManager;
  private connection?: IORedis | null;
  private isConnected: boolean = false;
  private isConnecting: boolean = false;
  private isDisabled: boolean = false;

  private constructor() {
    // Check if Redis is disabled
    this.isDisabled = process.env.DISABLE_REDIS_QUEUES === "true";

    if (this.isDisabled) {
      console.log("⚠️  Redis is disabled via DISABLE_REDIS_QUEUES=true");
      this.connection = null;
      return;
    }
  }

  public static getInstance(): RedisManager {
    if (!RedisManager.instance) {
      RedisManager.instance = new RedisManager();
    }
    return RedisManager.instance;
  }

  public async connect(): Promise<void> {
    if (this.isDisabled) {
      console.log("⚠️  Redis is disabled, skipping connection");
      return;
    }

    // If already connected, skip
    if (this.connection && this.isConnected) {
      console.log("📦 Using existing Redis connection");
      return;
    }

    // If currently connecting, wait for it to finish
    if (this.isConnecting) {
      console.log("⏳ Redis connection in progress, waiting...");
      // Wait for connection to complete
      await new Promise<void>((resolve) => {
        const checkInterval = setInterval(() => {
          if (!this.isConnecting) {
            clearInterval(checkInterval);
            resolve();
          }
        }, 100);
      });
      return;
    }

    this.isConnecting = true;

    try {
    const redisUrl = process.env.REDIS_URL;
    if (redisUrl) {
      this.connection = new IORedis(redisUrl, {
        maxRetriesPerRequest: null, // Required for BullMQ blocking operations
        enableAutoPipelining: true,
        lazyConnect: false,
        retryStrategy: (times: number) => {
        const delay = Math.min(times * 50, 2000);
        return delay;
        },
        reconnectOnError: (err) => {
        // Ignore SETINFO errors (compatibility issue with older Redis)
        if (err.message.includes("SETINFO")) {
          return false;
        }
        console.log("🔄 Redis reconnecting due to error:", err.message);
        return true;
        },
      });
    } else {
      this.connection = new IORedis({
        host: process.env.REDIS_HOST || "localhost",
        port: Number(process.env.REDIS_PORT) || 6379,
        password: process.env.REDIS_PASSWORD || undefined,
        maxRetriesPerRequest: null, // Required for BullMQ blocking operations
        enableAutoPipelining: true,
        lazyConnect: false,
        retryStrategy: (times: number) => {
        const delay = Math.min(times * 50, 2000);
        return delay;
        },
        reconnectOnError: (err) => {
        // Ignore SETINFO errors (compatibility issue with older Redis)
        if (err.message.includes("SETINFO")) {
          return false;
        }
        console.log("🔄 Redis reconnecting due to error:", err.message);
        return true;
        },
      });
    }

      this.setupEventHandlers();

      // IORedis connects automatically by default, no need to call connect()
      // Wait for ready event
      await new Promise<void>((resolve, reject) => {
        this.connection!.once("ready", () => {
          this.isConnected = true;
          this.isConnecting = false;
          resolve();
        });
        this.connection!.once("error", (err) => {
          // Ignore SETINFO errors
          if (err.message.includes("SETINFO")) {
            return;
          }
          this.isConnecting = false;
          reject(err);
        });
      });

      console.log("✅ Redis connected successfully");
    } catch (error) {
      console.error("❌ Redis connection failed:", error);
      this.connection = null;
      this.isConnected = false;
      this.isConnecting = false;
      throw error;
    }
  }

  public async disconnect(): Promise<void> {
    if (this.connection) {
      await this.connection.quit();
      this.isConnected = false;
      console.log("📦 Redis disconnected");
    }
  }

  public getConnection(): IORedis | null {
    return this.connection || null;
  }

  public isReady(): boolean {
    return this.isConnected && this.connection !== null && !this.isDisabled;
  }

  public async healthCheck(): Promise<boolean> {
    if (this.isDisabled || !this.connection) {
      return false;
    }

    try {
      await this.connection.ping();
      return true;
    } catch (error) {
      console.error("❌ Redis health check failed:", error);
      return false;
    }
  }

  private setupEventHandlers(): void {
    if (!this.connection) {
      return;
    }

    this.connection.on("connect", () => {
      console.log("✅ Redis connected");
    });

    this.connection.on("ready", () => {
      console.log("✅ Redis ready");
      this.isConnected = true;
    });

    this.connection.on("error", (err: any) => {
      // Ignore SETINFO errors (compatibility issue with older Redis versions)
      if (err.message.includes("SETINFO")) {
        return;
      }
      console.error("❌ Redis error:", err.message);
    });

    this.connection.on("close", () => {
      console.log("📦 Redis connection closed");
      this.isConnected = false;
    });

    this.connection.on("reconnecting", () => {
      console.log("🔄 Redis reconnecting...");
    });

    this.connection.on("end", () => {
      console.log("📦 Redis connection ended");
      this.isConnected = false;
    });
  }
}

export default RedisManager;
